mdp/nodes/sfa_nodes.py

```diff
@@ -369,6 +376,58 @@ def get_eta_values(self, t=1):
         return self._refcast(t / (2 * numx.pi) * numx.sqrt(self.d))
 
 
+class UnevenlySampledSFANode(SFANode):
+
+    def __init__(self, input_dim=None, output_dim=None, dtype=None,
+                 include_last_sample=True, rank_deficit_method='none'):
+        super(UnevenlySampledSFANode, self).__init__(
+            input_dim, output_dim, dtype, include_last_sample, rank_deficit_method)
+
+    def _init_cov(self):
+        # init two covariance matrices
+        # one for the input data
+        self._cov_mtx = UnevenlySampledCovarianceMatrix(self.dtype)
+        # one for the derivatives
+        self._dcov_mtx = UnevenlySampledCovarianceMatrix(self.dtype)
+
+    def time_derivative(self, x, dt=None):
+        """
+        Compute the linear approximation of the time derivative
+
+        :param x: The time series data.
+        :type x: numpy.ndarray
+
+        :returns: Piecewise linear approximation of the time derivative.
+        :rtype: numpy.ndarray
+        """
+        if dt is not None:
+            # Improvements can be made, by interpolating polynomials
+            return (x[1:, :]-x[:-1, :])/dt[:, None]
+        else:
+            # trivial fallback
+            return x[1:, :]-x[:-1, :]
+
+    def _train(self, x, dt, include_last_sample=None):
+        """
+        Training method.
+
+        :param x: The time series data.
+        :type x: numpy.ndarray
+
+        :param include_last_sample: For the ``include_last_sample`` switch have a
+            look at the SFANode.__init__ docstring.
+        :type include_last_sample: bool
+        """
+        if include_last_sample is None:
+            include_last_sample = self._include_last_sample
+        # works because x[:None] == x[:]
+        last_sample_index = None if include_last_sample else -1
+
+        # update the covariance matrices
+        self._cov_mtx.update(x[:last_sample_index, :], dt)
+        self._dcov_mtx.update(self.time_derivative(x, dt), dt[:-1])
```

Comment:
Please clarify how `dt` should be interpreted w.r.t. `include_last_sample`. In line 427, shouldn't it be `dt[:last_sample_index]`?

Reply:
Both should be shorter, I believe. What do you think should be the default behavior in multiple training phases with the trapezoid method? I think the chunks should be treated as separate integrals and added up.

Reply:
I assume you speak of "multiple chunks" when you say "multiple training phases", which are two things in MDP (SFA does not support or require multiple training phases in sense of multiple runs over the same data AFAIK).
I agree the integrals of multiple chunks should be added up, but ideally with an overlap of one frame (I think the dt of this particular frame shouldn't matter), i.e. such that the overall result is equal to computing one big chunk. I know this only works if the user provides the data with the right overlap. But *if* the user gets it right the node should yield the same result as one would get by using only a single chunk. This idea should IMO also form the way to test multi-chunk processing.

Reply:
In the most recent implementation I "linked" the chunks, essentially treating them as one integral. With optional specification of the inbetween timestep. If it is not supplied I take the mean of the bounding time steps. I also implemented a test, for checking, that the method returns the same as when supplied one big chunk. This however is not in line, if I understand correctly, with the current implementation of sfa described in the docstring of the init method. Using a the current integral approximation in sfa/covariance it does not matter, but the trapezoid method treats endpoints differently, thus it makes a difference.